examples/optim.jl

```diff
@@ -0,0 +1,299 @@
+    @assert size(migration_rates, 1) == size(migration_rates, 2) "migration_rates rates should be a square matrix"
+
+    C = length(Ns)
+    # normalize migration_rates
+    migration_rates_sum = sum(migration_rates, dims = 2)
+    for c in 1:C
+        migration_rates[c, :] ./= migration_rates_sum[c]
+    end
+
+    properties = @dict(
+        Ns,
+        Is,
+        β_und,
+        β_det,
+        β_det,
+        migration_rates,
+        infection_period,
+        infection_period,
+        reinfection_probability,
+        detection_time,
+        C,
+        death_rate
+    )
+    space = GraphSpace(complete_digraph(C))
+    model = ABM(PoorSoul, space; properties = properties)
+
+    # Add initial individuals
+    for city in 1:C, n in 1:Ns[city]
+        ind = add_agent!(city, model, 0, :S) # Susceptible
+    end
+    # add infected individuals
+    for city in 1:C
+        inds = get_node_contents(city, model)
+        for n in 1:Is[city]
+            agent = model[inds[n]]
+            agent.status = :I # Infected
+            agent.days_infected = 1
+        end
+    end
+    return model
+end
+
+function create_params(;
+  C,
+  Ns,
+  β_det,
+  migration_rate,
+  infection_period = 10,
+  reinfection_probability = 0.05,
+  detection_time = 3,
+  death_rate = 0.02,
+  Is = ones(Int, C),
+  β_und = [0.1 for i in 1:C],
+  )
+
+  migration_rates = reshape([migration_rate for i in 1:C*C], C,C)
+
+  params = @dict(
+      Ns,
+      β_und,
+      β_det,
+      migration_rates,
+      infection_period,
+      reinfection_probability,
+      detection_time,
+      death_rate,
+      Is
+  )
+
+  return params
+end
+
+function agent_step!(agent, model)
+  migrate!(agent, model)
+  transmit!(agent, model)
+  update!(agent, model)
+  recover_or_die!(agent, model)
+end
+
+function migrate!(agent, model)
+  nodeid = agent.pos
+  d = DiscreteNonParametric(1:(model.C), model.migration_rates[nodeid, :])
+  m = rand(d)
+  if m ≠ nodeid
+      move_agent!(agent, m, model)
+  end
+end
+
+function transmit!(agent, model)
+  agent.status == :S && return
+  rate = if agent.days_infected < model.detection_time
+      model.β_und[agent.pos]
+  else
+      model.β_det[agent.pos]
+  end
+
+  d = Poisson(rate)
+  n = rand(d)
+  n == 0 && return
+
+  for contactID in get_node_contents(agent, model)
+      contact = model[contactID]
+      if contact.status == :S ||
+         (contact.status == :R && rand() ≤ model.reinfection_probability)
+          contact.status = :I
+          n -= 1
+          n == 0 && return
+      end
+  end
+end
+
+update!(agent, model) = agent.status == :I && (agent.days_infected += 1)
+
+function recover_or_die!(agent, model)
+  if agent.days_infected ≥ model.infection_period
+      if rand() ≤ model.death_rate
+          kill_agent!(agent, model)
+      else
+          agent.status = :R
+          agent.days_infected = 0
+      end
+  end
+end
```

Comment:
Can some/all of this be replaced by a call to our new Model Library?

Reply:
All please. The focus of this docpage is well outside model creation.

Reply:
You had decided not to include the SIR model in the Model library.

Reply:
Ah! True!!!

Reply:
Let's chat about this tomorrow then perhaps.

Reply:
:( that is very unfortunate. How necessary is it for you @kavir1698 to use this graph based SIR model for such an example? I just feel like of having several pages of system definitions will make reading this page much harder than it has to. The main point is how to use BlackBoxOptim.

Reply:
Among the examples we have, this model is the best candidate for this analysis. Other ones either have too few parameters, or are just not that interesting. One solution is to put these definitions in a different file, and reference to that file?

Reply:
I second @kavir1698's analysis here @Datseris. Our other examples don't provide us with enough expressiveness.

Might be worth trying Ali. Take the [sir.jl](https://github.com/JuliaDynamics/Agents.jl/blob/409c359d2431dc2991076c0f497ee192259cf61c/src/models/sir.jl) file we dumped in #269 and import it with a single line. Something like:

```julia
include("sir.jl")
model, agent_step!, model_step! = sir()
```
Here we use the SIR model discussed in our [Examples documentation](https://juliadynamics.github.io/Agents.jl/stable/examples/sir/). For the complete implementation, see [`sir.jl`](./sir.jl).

Reply:
Sounds fair, but please replace the link above with `[SIR model for the spread of COVID-19](@ref)` so that the crosslink is generated automatically (and can work in any version of docs)

Reply:
I am going to to create a new file for this model because some functions are slightly different. The general logic is the same, so we can refer the reader to the examples documentation.